File: Mock Practice/quick_sort.py
import numpy as np
from numpy import array

class quicksort:
    def __init__(self):
        self.arr = array([1,9,8,11,6,17,4,-3,2,0])

    def partition(self,low,high):
        if low < high :
            i = low
            j = high
            pivot = self.arr[high]

            while (i<j):
                while (self.arr[i] < pivot):
                    if (i<j):
                        i = i + 1
                    else :
                        break
                
                while (self.arr[j] >= pivot):
                    j = j - 1 

                if (i<j):
                    self.arr[i],self.arr[j] = self.arr[j],self.arr[i]
            
            self.arr[i],self.arr[high] = self.arr[high],self.arr[i]
            
            return j 
    
    def quick_sort(self,low,high):
        if low < high:
            pi = self.partition(low,high)
            self.quick_sort(low,pi-1)
            self.quick_sort(pi+1,high)

q = quicksort()
q.quick_sort(0,9)
print(q.arr)

File: sortingalgo.cpp/quicksort.cpp
#include <iostream>
using namespace std;
void swap(int &a, int &b)
{
    int temp = a;
    a = b;
    b = temp;
}

void print_arr(int arr[], int n)
{
    for (int i = 0; i < n; i++)
    {
        cout << arr[i] << " ";
    }
    cout<<endl;
}

int partition (int arr[], int low, int high)
{
    if (low < high)
    {
        int i = low ;
        int j = high ;
        int pivot = arr[low];

        while (i<j)
        {
            while (arr[i] <= pivot)
            {
                if (i<high)
                {
                    i = i + 1;
                }
                else 
                {
                    break;
                }
            }

            while (arr[j] > pivot)
            {
                if(j>low)
                {
                    j = j - 1 ;
                }
                else 
                {
                    break;
                }
            }

            if (i<j)
            {
                swap(arr[i], arr[j]);
            }
            print_arr(arr,10);
            if (i==j)
            {
                cout<<low<<" "<<high<<endl;
                cout<<i<<j<<endl;
                print_arr(arr,10);
            }
        }
        swap(arr[low], arr[j]);
        
        return j ; 
    }
}

void quicksort(int arr[], int low, int high)
{
    if (low < high)
    {
        int pi = partition(arr, low, high);
        quicksort(arr,low, pi-1);
        quicksort(arr, pi+1, high);
    }
}


int main()
{
    int arr[] = {9, 10, 8, 6, 7, 5, 4, 3, 2, 1};
    quicksort(arr, 0,9);
    print_arr(arr, 10);

    return 0;
}